crates/payload/basic/src/lib.rs

```diff
@@ -967,6 +990,49 @@ fn commit_withdrawals<DB: Database<Error = Error>>(
     })
 }
 
+/// TODO: docs
+fn pre_block_contract_call<DB>(
+    db: State<DB>,
+    chain_spec: &ChainSpec,
+    block_number: u64,
+    initialized_cfg: &CfgEnv,
+    initialized_block_env: &BlockEnv,
+    attributes: &PayloadBuilderAttributes,
+) -> Result<State<DB>, PayloadBuilderError>
+where
+    DB: Database,
+    <DB as Database>::Error: Debug,
+{
+    // Configure the environment for the block.
+    let env = Env {
+        cfg: initialized_cfg.clone(),
+        block: initialized_block_env.clone(),
+        ..Default::default()
+    };
+
+    // apply pre-block EIP-4788 contract call
+    let mut evm_pre_block = revm::EVM::with_env(env);
+    evm_pre_block.database(db);
+
+    // initialize a block from the env, because the pre block call needs the block itself
+    match apply_pre_block_call(
+        chain_spec,
+        attributes.timestamp,
+        block_number,
+        attributes.parent_beacon_block_root,
+        &mut evm_pre_block,
```

Comment:
the txenv is still uninitialized here?

Reply:
Changed `fill_tx_env_with_beacon_root_contract_call` to explicitly set each field in the `TxEnv`